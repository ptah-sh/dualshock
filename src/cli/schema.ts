import { zodToJsonSchema } from "zod-to-json-schema";
import type { DualshockConfig } from "./config.js";

export const createSchema = (
	config: Pick<DualshockConfig["schema"], "registry" | "events">,
): any => {
	const schema = {
		$schemaVersion: "dualshock:1",
	} as any;

	schema.rpc = Object.entries(config.registry).reduce(
		(acc, [key, value]: [string, any]) => {
			const rpcSchema: any = {};

			if (value.args) {
				rpcSchema.args = zodToJsonSchema(value.args);
			}
			if (value.returns) {
				rpcSchema.returns = zodToJsonSchema(value.returns);
			}

			acc[key] = rpcSchema;

			return acc;
		},
		{} as any,
	);

	schema.emits = Object.entries(config.events).reduce(
		(acc, [key, value]: [string, any]) => {
			const eventSchema: any = {};
			if (value.payload) {
				eventSchema.payload = zodToJsonSchema(value.payload);
			}

			acc[key] = eventSchema;

			return acc;
		},
		{} as any,
	);

	return schema;
};
