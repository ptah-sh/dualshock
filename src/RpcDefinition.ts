import type { TypeOf, ZodType } from "zod";
import type { Refine } from "./refine";
import type { ContextInstance } from "./context";

export type RpcDefinition<
	A extends ZodType | undefined,
	R extends ZodType | undefined,
	Context extends object,
> = {
	name: string;
	args?: A;
    returns?: R;
	refine?: {
        args?: A extends ZodType ? Refine<TypeOf<A>, Context>[] : never;
        returns?: R extends ZodType ? Refine<TypeOf<R>, Context>[] : never;
    };
    fn: (args: A extends ZodType ? TypeOf<A> : undefined, context: ContextInstance<Context>) => Promise<R extends ZodType ? TypeOf<R> : void>;
}

/**
 * Simplifies the creation of a standalone RPC definition.
 *
 * Without this function you'd have to repeat the types for every RPC to get the
 * in-place type checking, and not in the server.rpc(myRpc) callsite.
 *
 * Usage:
 * ```ts
 * // myRpc.ts
 * export const myRpc = rpc({
 *    ...
 * })
 *
 * // rpcServer.ts
 * import { myRpc } from "./myRpc"
 * server.rpc(myRpc)
 * ```
 *
 * @param opts - The options for the RPC call
 * @returns The RPC definition with the provided options
 */
export const rpc = <A extends ZodType | undefined, R extends ZodType | undefined, Context extends object>(opts: RpcDefinition<A, R, Context>) => opts;
