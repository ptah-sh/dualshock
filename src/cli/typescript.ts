import { jsonSchemaToZod } from "json-schema-to-zod";

export const createTypescript = (args: {
	schema: { rpc: any; emits: any };
	rpcTypeName?: string;
	emitsTypeName?: string;
}): string => {
	const {
		schema,
		rpcTypeName = "DualshockInvokables",
		emitsTypeName = "DualshockEvents",
	} = args;

	const contents = [
		"// This file was automagically generated by the @ptah/dualshock package",
		"// DO NOT EDIT",
		"//",
	];

	contents.push("import { z, type ZodType } from 'zod';");
	contents.push("");

	contents.push(`export const ${emitsTypeName} = {`);
	for (const [event, { payload }] of Object.entries<any>(schema.emits)) {
		contents.push(`  "${event}": {`);
		contents.push(
			`    payload: ${jsonSchemaToZod(payload).replace(/\.strict\(\)/g, "")},`,
		);
		contents.push("  },");
	}
	contents.push("} as const;");
	contents.push("");

	contents.push(`export type ${emitsTypeName} = {`);
	contents.push(`  [key in keyof typeof ${emitsTypeName}]: {`);
	contents.push(
		`    payload: z.infer<typeof ${emitsTypeName}[key]["payload"]>;`,
	);
	contents.push("  };");
	contents.push("};");

	contents.push("");

	contents.push(`export const ${rpcTypeName} = {`);
	for (const [rpc, { args, returns }] of Object.entries<any>(schema.rpc)) {
		contents.push(`  "${rpc}": {`);
		contents.push(
			`    args: ${
				args ? jsonSchemaToZod(args).replace(/\.strict\(\)/g, "") : "undefined"
			},`,
		);
		contents.push(
			`    returns: ${
				returns
					? jsonSchemaToZod(returns).replace(/\.strict\(\)/g, "")
					: "undefined"
			},`,
		);
		contents.push("  },");
	}
	contents.push("} as const;");

	contents.push("");
	contents.push(`export type ${rpcTypeName} = {`);
	contents.push(`  [key in keyof typeof ${rpcTypeName}]: {`);
	contents.push(`    args: z.infer<typeof ${rpcTypeName}[key]["args"]>;`);
	contents.push(
		`    returns: ${rpcTypeName}[key]["returns"] extends ZodType<infer T> ? T : undefined;`,
	);
	contents.push("  };");
	contents.push("}");

	return contents.join("\n");
};
