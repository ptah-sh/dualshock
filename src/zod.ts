import type { ZodError } from "zod";

export function isZodError(err: unknown): err is ZodError {
	return err instanceof Error && err.constructor.name === "ZodError";
}
