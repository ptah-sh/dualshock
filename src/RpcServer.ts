import type { WebSocket, WebSocketServer } from "ws";
import { RpcRouter } from "./RpcRouter.js";
import type { Logger } from "pino";
import { RpcConnection } from "./RpcConnection.js";
import { WebSocketWs } from "./websocket/WebSocket.ws.js";
import { type TypeOf, type ZodTypeAny, z, type ZodType } from "zod";
import { BaseRpcClient } from "./BaseRpcClient.js";
import { rpc } from "./RpcDefinition.js";
import type { Plugin } from "./Plugin.js";

interface RpcServerOptions<
	Invokables extends Record<
		string,
		{ args: TypeOf<ZodTypeAny>; returns: TypeOf<ZodTypeAny> }
	>,
	Events extends Record<string, { payload: TypeOf<ZodTypeAny> }>,
> {
	logger: Logger;
	// TODO: move invokables and events under a single key - "clients"
	invokables: Invokables;
	events: Events;
}

export class RpcServer<
	Invokables extends Record<string, { args: ZodTypeAny; returns: ZodTypeAny }>,
	Events extends Record<string, { payload: ZodTypeAny }>,
> extends BaseRpcClient {
	constructor(args: RpcServerOptions<Invokables, Events>) {
		super(args.logger, new RpcRouter(args.logger));

		this.router.rpc(
			"ping",
			rpc()
				.args(z.date())
				.returns(z.date())
				.fn(async () => {
					return new Date();
				}),
		);
	}

	// TODO: add timeout for receiving the first message (handshake).
	// TODO: handlers should accept the WS connection / or it should be abstracted ???
	// TODO: add onConnection(RpcConnection) callback
	protected handleConnection(ws: WebSocket) {
		this.log.info("Incoming Connection");

		new RpcConnection<Invokables, Events>(
			new WebSocketWs(ws),
			this.log,
			this.router,
			{} as Invokables,
			{} as Events,
		);
	}

	async listen(wss: WebSocketServer): Promise<void> {
		wss.on("connection", this.handleConnection.bind(this));
		wss.on("error", this.handleError.bind(this));
		wss.on("close", this.handleClose.bind(this));
	}

	// TODO: create a PluginShim so that plugins can't manupulate server instances directly.
	// TODO: allow to pass plugin options
	use(plugin: Plugin<Invokables, Events>) {
		plugin.setup(this);
	}

	protected handleError(err: Error) {
		this.log.warn('RpcServer: "handleError"', err);
	}

	protected handleClose() {
		this.log.info("WebSocketServer connection closed");
	}
}
