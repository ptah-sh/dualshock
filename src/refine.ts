import { type ZodAny, z, type TypeOf, type ZodTypeAny } from "zod";

export type TRefine<
	A extends ZodTypeAny,
	R extends ZodTypeAny,
	C extends ZodTypeAny,
> = {
	path: (string | number)[];
	message: string;
	code?: string;
	context: C;
	args: A;
	pre?: (args: A, context: C) => Promise<boolean>;
	returns: R;
	post?: (args: A, returns: R, context: C) => Promise<boolean>;
};

export class RefineBuilder<
	A extends ZodTypeAny,
	R extends ZodTypeAny,
	C extends ZodTypeAny,
	TOmit extends keyof TRefine<A, R, C> | "schema" = "schema",
> {
	constructor(private build: TRefine<A, R, C>) {}

	path<B extends Omit<RefineBuilder<A, R, C, TOmit | "path">, TOmit | "path">>(
		path: (string | number)[],
	): B {
		this.build.path = path;

		return this as unknown as B;
	}

	message<
		B extends Omit<
			RefineBuilder<A, R, C, TOmit | "message">,
			TOmit | "message"
		>,
	>(message: string): B {
		this.build.message = message;

		return this as unknown as B;
	}

	code<B extends Omit<RefineBuilder<A, R, C, TOmit | "code">, TOmit | "code">>(
		code: string,
	): B {
		this.build.code = code;

		return this as unknown as B;
	}

	context<
		V extends C,
		B extends Omit<
			RefineBuilder<A, R, V, TOmit | "context">,
			TOmit | "context"
		>,
	>(context: V): B {
		this.build.context = context;

		return this as unknown as B;
	}

	args<
		V extends A,
		B extends Omit<RefineBuilder<V, R, C, TOmit | "args">, TOmit | "args">,
	>(args: V): B {
		this.build.args = args;

		return this as unknown as B;
	}

	returns<
		V extends R,
		B extends Omit<
			RefineBuilder<A, V, C, TOmit | "returns">,
			TOmit | "returns"
		>,
	>(returns: V): B {
		this.build.returns = returns;

		return this as unknown as B;
	}

	pre<
		B extends Omit<
			RefineBuilder<A, R, C, Exclude<TOmit, "refineSchema"> | "pre">,
			Exclude<TOmit, "refineSchema"> | "pre"
		>,
	>(pre: (args: TypeOf<A>, context: TypeOf<C>) => Promise<boolean>): B {
		this.build.pre = pre;

		return this as unknown as B;
	}

	post<
		B extends Omit<
			RefineBuilder<A, R, C, Exclude<TOmit, "refineSchema"> | "post">,
			Exclude<TOmit, "refineSchema"> | "post"
		>,
	>(
		post: (
			args: TypeOf<A>,
			returns: TypeOf<R>,
			context: TypeOf<C>,
		) => Promise<boolean>,
	): B {
		this.build.post = post;

		return this as unknown as B;
	}

	get refineSchema(): TRefine<any, any, any> {
		return this.build;
	}
}

export const refine = <
	A extends ZodTypeAny,
	R extends ZodTypeAny,
	C extends ZodTypeAny,
>(): RefineBuilder<A, R, C> => {
	return new RefineBuilder<A, R, C>({
		message: "Not Implemented",
		path: [],
		args: z.any() as unknown as A,
		returns: z.any() as unknown as R,
		context: z.any() as unknown as C,
	});
};
