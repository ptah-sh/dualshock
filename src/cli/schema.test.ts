import { describe, test, expect } from "vitest";
import { createSchema } from "./schema.js";
import { RpcServer } from "../RpcServer.js";
import { pino } from "pino";
import { rpc, on, emits } from "../RpcDefinition.js";
import { z } from "zod";
import { refine } from "../refine.js";
import { createTypescript } from "./typescript.js";

const server = new RpcServer({
	logger: pino({ enabled: false }),
	clients: {
		invokables: {},
		events: {},
	},
});

server
	.rpc(
		"rpc-in-root-ns",
		rpc()
			.args(z.number())
			.fn(async () => console.log("hello!")),
	)
	.ns("some-ns")
	.rpc(
		"rpc-in-some-ns",
		rpc()
			.context(z.object({}))
			.args(
				z.object({
					someArg: z.string(),
					nullableArg: z.number().nullable(),
				}),
			)
			.returns(z.string())
			.fn(async () => "hello!"),
	)
	.ns("sub-ns-in-some-ns")
	.on(
		"event-handler-in-sub-ns",
		on()
			.args(z.number())
			.refine([
				refine()
					.returns(z.string())
					.pre(async () => true),
			])
			.fn(async () => undefined),
	);

server.ns("some-ns-for-events").emits(
	"event-in-some-ns-for-events",
	emits().payload(
		z.object({
			eventKey: z.string(),
		}),
	),
);

describe("schema", () => {
	test("createSchema", () => {
		expect(
			createSchema({
				registry: server.registry(),
				events: server.events(),
			}),
		).toMatchSnapshot();
	});
});

describe("typescript", () => {
	test("createTypescript", () => {
		expect(
			createTypescript({
				schema: createSchema({
					registry: server.registry(),
					events: server.events(),
				}),
			}),
		).toMatchSnapshot();
	});
});
