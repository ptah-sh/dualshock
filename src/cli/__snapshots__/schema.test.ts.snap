// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema > createSchema 1`] = `
{
  "$schemaVersion": "dualshock:1",
  "emits": {
    "some-ns-for-events:event-in-some-ns-for-events": {
      "payload": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": false,
        "properties": {
          "eventKey": {
            "type": "string",
          },
        },
        "required": [
          "eventKey",
        ],
        "type": "object",
      },
    },
  },
  "rpc": {
    "ping": {
      "args": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "format": "date-time",
        "type": "string",
      },
      "returns": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "format": "date-time",
        "type": "string",
      },
    },
    "rpc-in-root-ns": {
      "args": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "type": "number",
      },
    },
    "some-ns:rpc-in-some-ns": {
      "args": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": false,
        "properties": {
          "nullableArg": {
            "type": [
              "number",
              "null",
            ],
          },
          "someArg": {
            "type": "string",
          },
        },
        "required": [
          "someArg",
          "nullableArg",
        ],
        "type": "object",
      },
      "returns": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "type": "string",
      },
    },
  },
}
`;

exports[`typescript > createTypescript 1`] = `
"// This file was automagically generated by the @ptah/dualshock package
// DO NOT EDIT
//
import { z, type ZodType } from 'zod';

export const DualshockEvents = {
  "some-ns-for-events:event-in-some-ns-for-events": {
    payload: z.object({ "eventKey": z.string() }),
  },
} as const;

export type DualshockEvents = {
  [key in keyof typeof DualshockEvents]: {
    payload: z.infer<typeof DualshockEvents[key]["payload"]>;
  };
};

export const DualshockInvokables = {
  "ping": {
    args: z.string().datetime({ offset: true }),
    returns: z.string().datetime({ offset: true }),
  },
  "rpc-in-root-ns": {
    args: z.number(),
    returns: undefined,
  },
  "some-ns:rpc-in-some-ns": {
    args: z.object({ "someArg": z.string(), "nullableArg": z.union([z.number(), z.null()]) }),
    returns: z.string(),
  },
} as const;

export type DualshockInvokables = {
  [key in keyof typeof DualshockInvokables]: {
    args: z.infer<typeof DualshockInvokables[key]["args"]>;
    returns: DualshockInvokables[key]["returns"] extends ZodType<infer T> ? T : undefined;
  };
}"
`;
